controllers/buildpipeline/buildpipeline_adapter.go

```diff
@@ -65,17 +65,28 @@ func NewAdapter(pipelineRun *tektonv1.PipelineRun, component *applicationapiv1al
 // to the build PipelineRun being processed exists. Otherwise, it will create a new pipeline Snapshot.
 func (a *Adapter) EnsureSnapshotExists() (controller.OperationResult, error) {
 	var err error
+	// a marker if we should remove finalizer from build PLR
+	var canRemoveFinalizer bool
+
 	defer func() {
 		err = tekton.AnnotateBuildPipelineRunWithCreateSnapshotAnnotation(a.context, a.pipelineRun, a.client, err)
 		if err != nil {
 			a.logger.Error(err, "Could not add create snapshot annotation to build pipelineRun", h.CreateSnapshotAnnotationName, a.pipelineRun)
 		}
+
+		if canRemoveFinalizer {
+			err := h.RemoveFinalizerFromPipelineRun(a.client, a.logger, a.context, a.pipelineRun, h.IntegrationPipelineRunFinalizer)
```

Comment:
because this is in defer function, if we fail to remove finalizer, it may stuck there because we won't retry

Reply:
I somehow feel that this defer function is not the right solution. I don't like the magic with global variables.

we should probably explicitly call the function to update annotation on places where it makes sense

Reply:
Yes, you are right. I realized it when I saw that annotation** in defer before, so I can't ignore it...
Maybe we have to remove the `defer`. Do you have any idea?

Reply:
Right now we're running into an increasing number of things we may need to do when the ensure functions exit.  We may want to have a general function that handles all parts of the cleanup that just gets called everywhere.  That way we're not dealing with 6 or 7 lines of repeated code everywhere we exit an Ensure function.

Reply:
I would still use defer but rewrite the returned operationResult